upgrade_v3.php

```diff
@@ -0,0 +1,1974 @@
+                writeToLog("All pre-upgrade checks passed successfully.");
+                $html .= "<h3>Ready to upgrade ✅</h3>
+                <br />Your system is compatible with Mautic 3!<br>Do not refresh or stop the process. This may take several minutes.<br><br>
+                " . $generalRemarks . "
+                <button class=\"btn btn-primary\" id=\"startUpgradeButton\">Start the upgrade</button>";
+            }
+
+            outputJSON('success', $html);
+
+        case 'startUpgrade':
+            writeToLog("Starting the upgrade...");
+            sendUpgradeStats('started');
+            outputJSON('success', 'OK', 'backupDatabase', 'Backing up the database if we can...');
+
+        case 'backupDatabase':
+            // Only do the backup if mysqldump is available, otherwise skip this step.
+            if (is_mysqldump_available() === true) {
+                writeToLog("Running database backup...");
+                list($success, $message) = backup_database();
+
+                if (!$success) {
+                    throwErrorAndWriteToLog(
+                        "ERR_DATABASE_BACKUP_FAILED",
+                        "Database backup failed. Your Mautic 2 installation is intact, so you can safely restart the upgrade. Error from mysqldump: " . $message
+                    );
+                }
+                writeToLog("Successfully backed up database.");
+            } else {
+                writeToLog("Skipping database backup due to mysqldump not being available.");
+            }
+
+            outputJSON('success', 'OK', 'applyV2Migrations', 'Applying Mautic 2 database migrations to ensure all migrations are in place prior to upgrade...');
+
+        case 'applyV2Migrations':
+            writeToLog("Getting the amount of available Mautic 2 database migrations...");
+            // Get the amount of available database migrations
+            list($success, $data) = get_available_migrations();
+
+            if ($success === false) {
+                throwErrorAndWriteToLog(
+                    "ERR_MAUTIC_2_MIGRATIONS_FAILED",
+                    'We couldn\'t reliably detect the amount of available database migrations. Please try again by refreshing this page.'
+                );
+            } else {
+                if (isset($data[2]) && isset($data[2][0])) {
+                    $availableMigrations = intval($data[2][0]);
+
+                    if ($availableMigrations > 0) {
+                        writeToLog("Applying Mautic 2 database migrations, " . $availableMigrations . " to go...");
+
+                        // Apply migrations on the 2.x branch just so we're sure that we have all migrations in place.
+                        list($success, $output) = apply_single_migration();
+
+                        if ($success === false) {
+                            throwErrorAndWriteToLog(
+                                "ERR_MAUTIC_2_MIGRATIONS_FAILED",
```

Comment:
I think this error code should be specific that the application of the migration has failed.

```suggestion
                                "ERR_MAUTIC_2_MIGRATIONS_APPLICATION_FAILED",
```

Reply:
Not necessary due to the introduction of the `_MIGRATIONS_IDENTIFICATION_FAILED` error codes in https://github.com/mautic/mautic/pull/8824/commits/2751b2787e2bad6c463ef59c056aae2ff281b251